tests/test_render.py

```diff
@@ -1866,3 +1881,94 @@ def update_output(*args):
+
+        app.config.suppress_callback_exceptions = True
+
+        app.layout = html.Div([
+            dcc.Tabs(
+                id="tabs",
+                children=[
+                    dcc.Tab(label="Tab 1", value="tab1", id="tab1"),
+                    dcc.Tab(label="Tab 2", value="tab2", id="tab2"),
+                ],
+                value="tab1",
+            ),
+
+            # Tab content
+            html.Div(id="tab_content"),
+        ])
+        tab1_layout = [
+            html.Div([dcc.Graph(id='graph1',
+                                figure={
+                                    'data': [{
+                                        'x': [1, 2, 3],
+                                        'y': [5, 10, 6],
+                                        'type': 'bar'
+                                        }]
+                                })]),
+
+            html.Pre(id='graph1_info'),
+        ]
+
+
+        tab2_layout = [
+            html.Div([dcc.Graph(id='graph2',
+                                figure={
+                                    'data': [{
+                                        'x': [4, 3, 2],
+                                        'y': [5, 10, 6],
+                                        'type': 'bar'
+                                        }]
+                                })]),
+
+            html.Pre(id='graph2_info'),
+        ]
+
+        @app.callback(Output(component_id='graph1_info', component_property='children'),
+                    [Input(component_id='graph1', component_property='clickData')])
+        def display_hover_data(hover_data):
+            print(hover_data)
+            show_well_info = json.dumps(hover_data)
+            return show_well_info
+
+
+        @app.callback(Output(component_id='graph2_info', component_property='children'),
+                    [Input(component_id='graph2', component_property='clickData')])
+        def display_hover_data(hover_data):
+            print(hover_data)
+            show_well_info = json.dumps(hover_data)
+            return show_well_info
+
+        @app.callback(Output("tab_content", "children"), [Input("tabs", "value")])
+        def render_content(tab):
+            if tab == "tab1":
+                return tab1_layout
+            elif tab == "tab2":
+                return tab2_layout
+            else:
+                return tab1_layout
+
+        self.startServer(app)
+
+        self.wait_for_element_by_css_selector('#graph1')
+
+        (self.driver.find_elements_by_css_selector(
+            '#graph1'
+        )[0]).click()
+
+        self.wait_for_text_to_not_equal('#graph1_info', 'null')
+
+        (self.driver.find_elements_by_css_selector(
+            '#tab2'
+        )[0]).click()
+
+        self.wait_for_element_by_css_selector('#graph2')
+
+        (self.driver.find_elements_by_css_selector(
+            '#graph2'
+        )[0]).click()
+        
+        self.wait_for_text_to_not_equal('#graph2_info', 'null')
```

Comment:
could you update your editor to always add a newline after saving a file? otherwise, the next person that makes a change to this file will have to commit the newline themselves

Reply:
Sure, feel free to add it to linting too. But note that this is very standard behaviour that we use across all projects.

Reply:
In other words, it's just a good default to have in your editor for all projects, whether they have linting in there or not: https://stackoverflow.com/a/5813359/4142536

Reply:
Well apparently it's not a default in VSCode. I'm fine with toggling that switch ofcourse, just saying it might be easier to enforce those rules in linting if you feel strongly about them and if it is considered standard behaviour for all of our projects.

Reply:
Yup, go ahead and add it to eslint. Just note that not all projects have linting, and so I recommend making it your default in your code editor as well as per our organizational standard practices.

Reply:
Sure, will make it my default. Do you have a link for me for the organizational standard practices? I'm looking through our dev docs but can't find anything about that.

Reply:
For python you should have atleast flake8.

Reply:
@valentijnnieman here are some standard practice docs https://github.com/plotly/dev-docs/blob/master/reference/styleguides/python.md.
pep8 does recommend a newline at the end of a file but as @chriddyp suggested it is fairly standard programming practice (across languages and linters) to include a newline terminator https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline. 

So while it _should_ be in a linter it is better to turn it on in your editor options so that you append a newline even in projects that don't have linting.